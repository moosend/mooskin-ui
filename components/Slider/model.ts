import { IInputCallbackData } from '../_utils/types/commonTypes';
import { IInputComponentProps } from '../Input/model';

export interface ISliderComponentProps extends IInputComponentProps {
	/** what data is being used, helps whn extracting user input, you know on what field changes are made */
	dataLabel?: string;

	/** callback that is called when the Slider changes */
	onChangeSlider?: (e: React.ChangeEvent<HTMLInputElement>, data: IInputCallbackData) => void;
}
