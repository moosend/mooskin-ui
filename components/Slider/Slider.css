@import url('https://fonts.googleapis.com/css?family=Alegreya+Sans:500|Convergence');

.sliderContainer{
    z-index: 0;
    display: flex;
    flex-direction: column;
    position: relative;
}

.sliderContainer input[type=range], .sliderContainer input[type=range]::-webkit-slider-thumb {
    appearance: none;
    background: transparent;
}

.slider {
    flex-grow: 1;
    flex-shrink: 1;
    cursor: pointer;
    opacity: 1;
    transition: all 0.2s;
}

.slider:disabled {
    opacity: 1;
}

.slider:hover{
    opacity: 1;
}

.disabled {
    cursor: not-allowed;
}

.sliderLabel {
    color: #6b6b6b;
     /* font-size: 13px;  */
     font-size: .92857142857rem; 
     margin-bottom: 5px;
}

.trackLabels{
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    margin-top: 5px;
}

.trackLabel {
    color: #6b6b6b;
     /* font-size: 13px;  */
     font-size: .75rem; 
}

.tooltip{
    position: absolute;
    border-radius: 4px;
    padding: 5px;
    border: 1px solid #eaeaea;
    top: -20px;
    left: 0px;
    z-index: 1;
    background: #ffffff;
    font-size: .875rem;
    font-family: 'Convergence', 'Alegreya Sans', sans-serif;
}

.sliderContainer input[type=range]::-webkit-slider-thumb {
    appearance: none;
    border: 1px solid #d3d3d3;
    height: 20px;
    width: 20px;
    border-radius: 4px;
    background: #efefef;
    cursor: pointer;
    margin-top: -4px;
    box-shadow: 1px 1px 1px #aaa, 0px 0px 1px #aaa;
}

.sliderContainer input[type=range]::-ms-thumb {
    appearance: none;
    border: 1px solid #d3d3d3;
    height: 20px;
    width: 20px;
    border-radius: 4px;
    background: #efefef;
    cursor: pointer;
    margin-top: -4px;
    box-shadow: 1px 1px 1px #aaa, 0px 0px 1px #aaa;
}

.sliderContainer input[type=range]::-ms-tooltip {
    display: none;
}

.sliderContainer input[type=range]:active::-webkit-slider-thumb {
    background: #ffffff;
}

.sliderContainer input[type=range]:disabled::-webkit-slider-thumb {
    background: #dbdbdb;
    opacity: 1;
    cursor: not-allowed;
}

.sliderContainer input[type=range]::-webkit-slider-runnable-track {
    appearance: none;
    height: 13px;
    cursor: pointer;
    box-shadow: 1px 1px 1px #aaa, 0px 0px 1px #aaa;
    /* background: transparent; */
    border-radius: 1.3px;
    border: 0.2px solid #aaa;
}

.sliderContainer input[type=range]:disabled::-webkit-slider-runnable-track {
    appearance: none;
    background: #e9e9e9;
    cursor: not-allowed;
    opacity: 1;
}

.sliderContainer input[type=range]::-moz-range-track {
    appearance: none;
    height: 13px;
    cursor: pointer;
    box-shadow: 1px 1px 1px #aaa, 0px 0px 1px #aaa;
    /* background: transparent; */
    border-radius: 1.3px;
    border: 1px solid #aaa;
}

.sliderContainer input[type=range]::-ms-track {
    appearance: none;
    height: 13px;
    cursor: pointer;
    /* background: transparent; */
    border-color: transparent;
    border-width: 16px 0;
    color: transparent;

}

.classesIE {
    position: absolute;
    width: 100%;
    top: 0px;
    left: 0px;
    bottom: 0px;
    right: 0px;
}

.fill {
    height: 13px;
    width: 0px;
    cursor: pointer;
    background: #eeeeee;
    box-shadow: 1px 1px 1px #aaa, 0px 0px 1px #aaa;
    /* background: transparent; */
    border-radius: 3px;
    border: 0.2px solid #aaa;
    position: absolute;
    top: 2px;
    left: 2px;
    z-index: -1;
}

.disabledFill{
    background: #cccccc;
}

.edgeClasses {
    width: 100%;
}

input.edgeClasses[type=range]::-ms-fill-lower {
    background: #eeeeee;
    border: 0.2px solid #aaa;
    border-radius: 2.6px;
    box-shadow: 1px 1px 1px #aaa, 0px 0px 1px #aaa;
}

input.edgeClasses[type=range]::-ms-fill-upper {
    background: #fff;
    box-shadow: 1px 1px 1px #aaa, 0px 0px 1px #aaa;
    /* background: transparent; */
    border-radius: 1.3px;
    border: 0.2px solid #aaa;
}

.abContainer {
    z-index: 0;
    display: flex;
    flex-direction: column;
    position: relative;
    height: 100px;
}

.abContainer input[type=range]::-webkit-slider-thumb {
    appearance: none;
    border: 1px solid #d3d3d3;
    height: 110px;
    width: 20px;
    border-radius: 4px;
    background: #efefef;
    cursor: pointer;
    margin-top: -4px;
    box-shadow: 1px 1px 1px #aaa, 0px 0px 1px #aaa;
    background: green;
}

.abContainer input[type=range], .abContainer input[type=range]::-webkit-slider-thumb {
    appearance: none;
    background: transparent;
}

.abContainer input[type=range]::-webkit-slider-thumb {
    appearance: none;
    border: 1px solid #d3d3d3;
    height: 110px;
    width: 10px;
    border-radius: 4px;
    background: #efefef;
    cursor: pointer;
    margin-top: 0px;
    box-shadow: 1px 1px 1px #aaa, 0px 0px 1px #aaa;
}

.abContainer input[type=range]::-webkit-slider-thumb:hover {
    cursor: col-resize;
}

.abContainer input[type=range]:active::-webkit-slider-thumb {
    background: #ffffff;
}

.abContainer input[type=range]::-webkit-slider-runnable-track {
    appearance: none;
}

.abContainer input[type=range]:disabled::-webkit-slider-runnable-track {
    appearance: none;
}

.abContainer input[type=range]::-moz-range-track {
    appearance: none;
}

.abContainer input[type=range]::-ms-track {
    appearance: none;
    visibility: hidden;
    /* position: absolute;
    width: 100%;
    top: 0px;
    left: 0px;
    bottom: 0px;
    right: 0px; */
}

.abContainer input[type=range] {
    position: absolute;
    width: 100%;
    top: 0px;
    left: 0px;
    bottom: 0px;
    right: 0px;
    background-color: transparent;
    color: transparent;
}

.abContainer input[type=range]::-ms-fill-lower {
    background-color: transparent;
    color: transparent;
}


.abContainer input[type=range]::-ms-thumb {
    appearance: none;
    visibility: hidden;
}

.section {
    position: absolute;
    display: flex;
    flex-direction: column;
    justify-content: space-evenly;
    align-items: center;
    z-index: -1;
    top: 0px;
    bottom: 0px;
    text-align: center;
}

.font {
    font-size: 32px;
    font-weight: 400;
    color: #6b6b6b;
    font-family: 'Convergence', 'Alegreya Sans', sans-serif;
}

.abContainer input:hover {
    cursor: col-resize;
}

.sectionA {
    border-top-left-radius: 10px;
    border-bottom-left-radius: 10px;
    background: #5ccddf;
}

.sectionB {
    background: #f48770;
}

.sectionW {
    background: #f2c14a;
    border-top-right-radius: 10px;
    border-bottom-right-radius: 10px;
}

.abSlider{
    position: absolute;
    top: 0;
    right: 0;
    left: 0;
    bottom: 0;
    width: 100%;
    margin: 0;
}

.percentText {
    color: #6b6b6b;
    font-size: 14px;
}