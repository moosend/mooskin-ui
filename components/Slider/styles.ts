import styled from 'styled-components';

// Models
import { ISliderComponentProps } from './model';

// Components
import { Box } from '../Box/Box';

export const StyledSlider = styled(Box)<ISliderComponentProps>`
	-webkit-appearance: none;
	width: 100%;
	background: transparent;
	cursor: pointer;

	:focus {
		outline: none;
	}
	::-webkit-slider-runnable-track {
		border-radius: 20px;
		border: solid 1px #e2e2e2;
		background-color: #e7e7e7;
		height: 12px;
	}
	::-webkit-slider-thumb {
		box-shadow: 0 3px 6px 0 rgba(0, 0, 0, 0.16);
		border: solid 1px #9d9d9d;
		width: 33px;
		height: 33px;
		border-radius: 3px;
		background-color: #ffffff;
		border-radius: 50%;
		-webkit-appearance: none;
		margin-top: -11px;
	}
	::-moz-range-track {
		border-radius: 20px;
		border: solid 1px #e2e2e2;
		background-color: #e7e7e7;
		height: 12px;
	}
	::-moz-range-thumb {
		box-shadow: 0 3px 6px 0 rgba(0, 0, 0, 0.16);
		border: solid 1px #9d9d9d;
		width: 33px;
		height: 33px;
		border-radius: 3px;
		background-color: #ffffff;
		border-radius: 50%;
	}
	::-ms-track {
		border-radius: 20px;
		border: solid 1px #e2e2e2;
		background-color: #e7e7e7;
		height: 12px;
	}
	::-ms-thumb {
		box-shadow: 0 3px 6px 0 rgba(0, 0, 0, 0.16);
		border: solid 1px #9d9d9d;
		width: 33px;
		height: 33px;
		border-radius: 3px;
		background-color: #ffffff;
		border-radius: 50%;
	}
`;

StyledSlider.displayName = 'StyledSlider';
