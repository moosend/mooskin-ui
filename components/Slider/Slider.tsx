import * as React from 'react';

import styles from './Slider.css';

// interface IDocumentProps {
//     documentMode?: any;
//     StyleMedia?: any;
// }

import {IInputCallbackData} from '../_utils/types/commonTypes';

export interface ISliderProps{

    /** id of the component */
    id?: string;

    /** override Slider classes */
    className?: string;

    /** override Slider styles */
    style?: React.CSSProperties;

    /** wether the Slider is disabled */
    disabled?: boolean;

    /** Slider label */
    label?: string;

    /** adds track labels to the Slider */
    trackLabels?: boolean;

    /** adds current value tooltip to the Slider on sliding */
    tooltip?: boolean;

    /** value of the Slider */
    value: string | number;

    /** fill className */
    fillClassName?: string;

    /** styles for fill */
    fillStyle?: React.CSSProperties;

    /** range of the Slider */
    range?: number[];

    /** custom values for the Slider */
    values?: Array<number | string>;

    /** what data is being used, helps whn extracting user input, you know on what field changes are made */
    dataLabel?: string;

    /** callback that is called when the Slider changes */
    onChange?: (e: React.ChangeEvent<HTMLInputElement>, data: IInputCallbackData) => void;
}

export interface ISliderState {
    dragging: boolean;
    left: number;
    // fill: any;
}

export default class Slider extends React.Component<ISliderProps, ISliderState>{

    static displayName = 'Slider';

    slider: any;

    constructor(props: ISliderProps){
        super(props);

        this.state = {
            dragging: false,
            // fill: <div/>,
            left: 0
        };
    }

    componentDidMount(){
        this.forceUpdate();
    }

    // componentWillReceiveProps(){
    //     this.renderFill();
    // }

    render(){

        const {label, tooltip, trackLabels} = this.props;

        const slider = this.renderSlider();

        return (
            <div className={`${styles.sliderContainer}`}>
                {label && <label className={styles.sliderLabel} >{label}</label>}
                {slider}
                {!this.isEdge() && this.renderFill()}
                {trackLabels && this.renderTrackLabels()}
                {tooltip && this.renderTooltip()}
            </div>
        );
    }

    renderSlider = () => {
        if (this.props.range){
            // return this.rangeSlider();
            const min = this.props.range ? this.props.range[0] : 1;
            const max = this.props.range ? this.props.range[1] : 100;
            return this.getSliderWithType(min, max, this.props.value as number, this.onSliderChange);
        } else if (this.props.values) {
            const min = 0;
            const max = this.props.values ? this.props.values.length - 1 : 100;
            const value = this.props.values && this.props.values.indexOf(this.props.value);
            return this.getSliderWithType(min, max, value, this.onValuesSliderChange);
        }
    }

    getSliderWithType = (min: number, max: number, value: number, callback: any) => {

        const disabled = this.props.disabled ? styles.disabled : '';
        const edgeClasses = this.isEdge() ? styles.edgeClasses : '';

        const classesIE = this.isIE() ? styles.classesIE : '';

        return (
            <input
                ref={(slider) => this.slider = slider}
                id={this.props.id}
                style={this.props.style}
                className={`${styles.slider} ${disabled} ${edgeClasses} ${classesIE} ${this.props.className}`}
                value={value}
                min={min}
                max={max}
                type="range"
                onChange={callback}
                disabled={this.props.disabled}
                onMouseDown={this.onMouseDown}
                onMouseUp={this.onMouseUp}
                onMouseMove={this.onMouseMove}
            />
        );
    }

    renderTrackLabels = () => {

        const {range, values} = this.props;
        let type: any[] = [];

        if (range){
            type = range;
        } else if (values){
            type = values;
        } else {
            return '';
        }

        const labels = type.map((label, i) => {
            return <span key={i} className={styles.trackLabel}>{label}</span>;
        });

        return (
            <div className={styles.trackLabels}>
                {labels}
            </div>
        );
    }

    renderTooltip = () => {
        const display = this.state.dragging ? {} : {display: 'none'};
        const top = this.props.label ? {} : {top: -40};
        return (
            <div
                style={{...display, left: this.state.left, ...top}}
                className={styles.tooltip}
            >
                {this.props.value}
            </div>
        );
    }

    onSliderChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        this.props.onChange &&
        this.props.onChange(e, {value: e.target.value, dataLabel: this.props.dataLabel});
    }

    onValuesSliderChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        const value = this.props.values && this.props.values[e.target.value as any];
        this.props.onChange &&
        this.props.onChange(e, {value, dataLabel: this.props.dataLabel});
    }

    onMouseMove = (e: React.MouseEvent<HTMLInputElement>) => {
        // const mouseStart = window.innerWidth - e.pageX;
        const {dragging} = this.state;
        if (dragging) {
            // console.log(e.nativeEvent.offsetX);
            const left = this.getPosition(e);
            this.setState({left});
        }
    }

    getPosition = (e: React.MouseEvent<HTMLInputElement>) => {
        // let pos = e.nativeEvent.offsetX;
        const sliderWidth = this.slider.offsetWidth - 25;
        if (e.nativeEvent.offsetX < 0){
            return 0;
        } else if (e.nativeEvent.offsetX > sliderWidth){
            return sliderWidth;
        } else {
            return e.nativeEvent.offsetX;
        }
    }

    getStepsCount = () => {
        const start = this.props.range && this.props.range[0] || 0;
        const end = this.props.range && this.props.range[1] || 0;
        const steps = [];
        for (let i = start; i <= end; i++){
            steps.push(i);
        }
        return steps.indexOf(parseInt(this.props.value.toString(), 10));
    }

    getThumbPosition = () => {
        // const width = this.slider ? this.slider.value / this.slider.max : 50;
        const max = this.props.values ? this.slider.max : this.props.range ?
        this.props.range[1] - this.props.range[0] : 0;
        const value = this.props.range ? this.getStepsCount() : this.slider.value;
        const pos = value / max;
        const thumbPos = this.slider.clientWidth * pos;
        const width = 100 / (this.slider.clientWidth / thumbPos);
        return width > 98 ? 98 : width;
    }

    renderFill = () => {
        const {fillClassName, fillStyle} = this.props;
        const width = this.slider ? this.getThumbPosition() + '%' : '0%';
        const top = this.props.label ? {top: '23px'} : {};
        const disabled = this.props.disabled ? styles.disabledFill : '';
        return (
            <div className={`${styles.fill} ${disabled} ${fillClassName}`} style={{width, ...top, ...fillStyle}} />
        );
        // this.setState({fill});
    }

    onMouseDown = (e: React.MouseEvent<HTMLInputElement>) => {
        const left = this.getPosition(e);
        this.setState({dragging: true, left});
    }

    onMouseUp = (e: React.MouseEvent<HTMLInputElement>) => {
        this.setState({dragging: false});
    }

    isEdge = () => {
        // tslint:disable-next-line
        const win = window as any;
        const isIE = this.isIE();
        const isEdge = !isIE && !!win.StyleMedia;
        return isEdge;
    }

    isIE = () => {
        // tslint:disable-next-line
        const doc = document as any;
        const isIE = false || !!doc.documentMode as any;
        return isIE;
    }

}

export interface IABSliderProps {
    id?: string;
    dataLabel?: string;
    percentage: number;
    count: number;
    max?: number;
    min?: number;
    style?: React.CSSProperties;
    className?: string;
    onChange?: (e: React.ChangeEvent<HTMLInputElement>, data: IInputCallbackData) => void;
}

export const ABSlider: React.StatelessComponent<IABSliderProps> = (props) => {

    const min = props.min ? props.min : 5;
    const max = props.max ? props.max : 40;

    const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        const value = parseInt(e.target.value, 10) / 2 < min ? min :
                        parseInt(e.target.value, 10) / 2 > max ? max :
                        parseInt(e.target.value, 10) / 2;
        props.onChange &&
        props.onChange(e, {value, dataLabel: props.dataLabel});
    };

    const getPercentage = (num: number, per: number) => {
        const value = Math.floor((num / 100) * per);
        return value < 1 ? 1 : value;
    };

    const getWinnerPercentage = (num: number, per: number) => {
        return Math.ceil((num / 100) * per);
    };

    const percentage = props.percentage > max ? max : props.percentage < min ? min : props.percentage;

    const percentageAB = getPercentage(props.count, percentage);
    // const percentageWinner = getPercentage(props.count, 100 - percentage * 2);
    const percentageWinner = props.count <= 1000 ? props.count - 2 * percentageAB : getWinnerPercentage(props.count, 100 - percentage * 2);

    return(
        <div className={`${styles.abContainer} ${props.className}`} style={props.style}>
            <div
                className={`${styles.section} ${styles.sectionA}`}
                style={{left: 0, right: `${100 - percentage}%`}}
            >
                <span className={styles.font}>A</span>
                <span className={styles.percentText}>{percentage}% ({props.count === 0 ? 0 : percentageAB})</span>
            </div>
            <div
                className={`${styles.section} ${styles.sectionB}`}
                style={{left: `${percentage}%`, right: `${100 - 2 * percentage}%`}}
            >
                <span className={styles.font}>B</span>
                <span className={styles.percentText}>{percentage}% ({props.count === 1 || props.count === 0 ? 0 : percentageAB})</span>
            </div>
            <div
                className={`${styles.section} ${styles.sectionW}`}
                style={{left: `${2 * percentage}%`, right: 0}}
            >
                <span className={styles.font}>Winner</span>
                <span className={styles.percentText}>
                    {100 - percentage * 2}% ({percentageWinner < 0 ? 0 : percentageWinner})
                </span>
            </div>
            <input
                id={props.id}
                value={props.percentage * 2}
                type="range"
                style={{margin: 0}}
                step={2}
                min={0}
                max={100}
                onChange={onChange}
            />
        </div>
    );
};

ABSlider.displayName = 'ABSlider';
