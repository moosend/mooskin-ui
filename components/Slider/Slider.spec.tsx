import * as React from 'react';
import { Slider } from './Slider';

import { mount } from 'enzyme';

describe('Slider', () => {
	test('renders correctly', () => {
		const tree = mount(<Slider value={5} min={0} max={10} />);
		expect(tree).toMatchSnapshot();
	});
});
