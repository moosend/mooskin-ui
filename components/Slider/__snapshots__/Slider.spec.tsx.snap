// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider render AB Slider correctly 1`] = `
<ABSlider
  count={1370}
  onChange={[Function]}
  percentage={20}
>
  <div
    className="abContainer undefined"
  >
    <div
      className="section sectionA"
      style={
        Object {
          "left": 0,
          "right": "80%",
        }
      }
    >
      <span
        className="font"
      >
        A
      </span>
      <span
        className="percentText"
      >
        20
        % (
        274
        )
      </span>
    </div>
    <div
      className="section sectionB"
      style={
        Object {
          "left": "20%",
          "right": "60%",
        }
      }
    >
      <span
        className="font"
      >
        B
      </span>
      <span
        className="percentText"
      >
        20
        % (
        274
        )
      </span>
    </div>
    <div
      className="section sectionW"
      style={
        Object {
          "left": "40%",
          "right": 0,
        }
      }
    >
      <span
        className="font"
      >
        Winner
      </span>
      <span
        className="percentText"
      >
        60
        % (
        822
        )
      </span>
    </div>
    <input
      max={100}
      min={0}
      onChange={[Function]}
      step={2}
      style={
        Object {
          "margin": 0,
        }
      }
      type="range"
      value={40}
    />
  </div>
</ABSlider>
`;

exports[`Slider renders correctly 1`] = `
<div
  className="sliderContainer"
>
  <label
    className="sliderLabel"
  >
    Slider Label
  </label>
  <input
    className="slider disabled   myClass"
    disabled={true}
    id="Slider1"
    max={10}
    min={0}
    onChange={[Function]}
    onMouseDown={[Function]}
    onMouseMove={[Function]}
    onMouseUp={[Function]}
    style={
      Object {
        "color": "blue",
      }
    }
    type="range"
    value={4}
  />
  <div
    className="fill disabledFill undefined"
    style={
      Object {
        "top": "23px",
        "width": "0%",
      }
    }
  />
  <div
    className="trackLabels"
  >
    <span
      className="trackLabel"
      key="0"
    >
      0
    </span>
    <span
      className="trackLabel"
      key="1"
    >
      10
    </span>
  </div>
  <div
    className="tooltip"
    style={
      Object {
        "display": "none",
        "left": 0,
      }
    }
  >
    4
  </div>
</div>
`;
